File: frontend/src/utils/allocationEstimates.ts
import type { AllocationResponse } from '../types/api';

/** Зеркало backend `effective_estimate_hours`:
 *  если хоть одно override_estimate_*_hours не null → берём 4 цифры из override
 *  (null → 0); иначе — estimate_*_hours от BacklogItem. */
export function effectiveEstimate(a: AllocationResponse): {
  analyst: number;
  dev: number;
  qa: number;
  opo: number;
} {
  const hasOverride =
    a.override_estimate_analyst_hours !== null ||
    a.override_estimate_dev_hours !== null ||
    a.override_estimate_qa_hours !== null ||
    a.override_estimate_opo_hours !== null;
  if (hasOverride) {
    return {
      analyst: a.override_estimate_analyst_hours ?? 0,
      dev: a.override_estimate_dev_hours ?? 0,
      qa: a.override_estimate_qa_hours ?? 0,
      opo: a.override_estimate_opo_hours ?? 0,
    };
  }
  return {
    analyst: a.estimate_analyst_hours ?? 0,
    dev: a.estimate_dev_hours ?? 0,
    qa: a.estimate_qa_hours ?? 0,
    opo: a.estimate_opo_hours ?? 0,
  };
}

File: app/services/allocation_estimates.py
"""Effective per-role часы allocation: override если задан, иначе BacklogItem."""
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.models import ScenarioAllocation


_OVERRIDE_FIELDS = (
    "override_estimate_analyst_hours",
    "override_estimate_dev_hours",
    "override_estimate_qa_hours",
    "override_estimate_opo_hours",
)


def has_override(allocation: "ScenarioAllocation") -> bool:
    """True если хотя бы одно из 4 override-полей не NULL."""
    return any(getattr(allocation, f) is not None for f in _OVERRIDE_FIELDS)


def effective_estimate_hours(allocation: "ScenarioAllocation") -> dict[str, float]:
    """Per-role часы allocation: override приоритетнее BacklogItem.

    Правило inherit:
      - has_override(allocation) is True  → берём 4 цифры из override (NULL → 0.0)
      - has_override(allocation) is False → берём из allocation.backlog_item.estimate_*
    """
    if has_override(allocation):
        return {
            "analyst": float(allocation.override_estimate_analyst_hours or 0.0),
            "dev": float(allocation.override_estimate_dev_hours or 0.0),
            "qa": float(allocation.override_estimate_qa_hours or 0.0),
            "opo": float(allocation.override_estimate_opo_hours or 0.0),
        }
    bi = allocation.backlog_item
    return {
        "analyst": float(getattr(bi, "estimate_analyst_hours", 0) or 0.0),
        "dev": float(getattr(bi, "estimate_dev_hours", 0) or 0.0),
        "qa": float(getattr(bi, "estimate_qa_hours", 0) or 0.0),
        "opo": float(getattr(bi, "estimate_opo_hours", 0) or 0.0),
    }
